README.md
# SWTOR.app

Star Wars: The Old Republic app for macOS, working out of the box.

## How to install?

1. Download SWTOR.dmg from the [latest release](https://github.com/imanel/swtor-app/releases)
2. Open it
3. Move SWTOR.app to /Applications folder
4. Run like any other app

## FAQ:

**I see a weird message for a second during first time running the app, what should I do?**

```
The Wine configuration in /Applications/SWTOR.app/Contents/Resources/prefix is being updated, please wait...
```

Please ignore this message, this is popping only during first boot and should not appear any further.
It's a side effect of the way the application is built and have no negative effects.
In theory it could be avoided by running application once before packaging it and providing to you, but that could
resoult in even different side effects, so it was left alone.

**Just after launching the first time the launcher updates itself several times, is that normal?**

Yes, the launcher that is downloadable is known to be quite old and with every new installation it will go into
update-restart cycle for several times. This should be a one-time case, and in future it will update only when needed.

**After logging in I see "PLAY" button active, but I cannot press it**

Please wait a moment - for some reason it takes couple seconds for launcher to figure out that it need to download
full game (or future update), it should start updating automatically soon.

**Application is not starting/application crashed**

Please create [an issue](https://github.com/imanel/swtor-app/issues) and provide your Mac model, macOS version,
and content of `~/Library/Logs/SWTOR.log` file.

**After pressing "play" it takes ages to load game for the first time**

This might be the result of SWTOR enabling ultra quality and the highest possible resolution. It might take a while,
but as long as you can see spinner it will finally load. You can also press CMD+TAB to minimize the application,
and wait couple minutes while doing something else - it will keep loading in the background, and the next time
you will bring it back to front it should load correctly. After adjusting graphic settings it should boot pretty
much instantly the next time.

## How to build it by myself?

1. Make sure that you have homebrew installed
2. Check that you don't have any other Wine version installed (we're using [wine-crossover](https://github.com/Gcenx/homebrew-wine))
3. Run `./build.sh` to create Wine Prefix in `prefix` folder
4. Run `./package.sh` to create SWTOR.app and package it
5. You can find package in `dist` folder

## Acknowledgment

This app would not be possible without [excellend script written by AgentRG](https://github.com/AgentRG/swtor_on_mac),
as well as [pre-compiled Crossover Wine provided by Gcenx](https://github.com/Gcenx/homebrew-wine).

## Support

If you like my work then consider supporting me:

[![Donate with Bitcoin](https://en.cryptobadges.io/badge/small/bc1qmxfc703ezscvd4qv0dvp7hwy7vc4kl6currs5e)](https://en.cryptobadges.io/donate/bc1qmxfc703ezscvd4qv0dvp7hwy7vc4kl6currs5e)

[![Donate with Ethereum](https://en.cryptobadges.io/badge/small/0xA7048d5F866e2c3206DC95ebFa988fF987c0BccB)](https://en.cryptobadges.io/donate/0xA7048d5F866e2c3206DC95ebFa988fF987c0BccB)


CHANGELOG.md
# Changelog

## Latest

- switch to Crossover Wine 22 style

## 1.0.2

- cleanup and better packaging

## 1.0.1

- fix for Intel-based Macs using SWTOR-light version (no impact on SWTOR-standalone)

## 1.0.0

- initial release


build.sh
#!/usr/bin/env arch -x86_64 bash

set -e

SCRIPT_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
PREFIX_DIR="$SCRIPT_DIR/prefix"

SWTOR_DOWNLOAD=http://www.swtor.com/download

log() {
  echo -e "\033[01;35m\t$1\033[00m"
}

install_package_wget() {
  log "(1/4) Installing wget"
  brew install wget
}

tap_into_gcenx_brew() {
  log "(2/4) Tap into Gcenx/homebrew-wine"
  brew tap Gcenx/homebrew-wine
}

install_package_wine_stable() {
  log "(3/4) Installing latest Wine version"
  brew update
  brew install --cask --no-quarantine wine-crossover
}

install_package_winetricks() {
  log "(4/4) Installing Winetricks"
  brew install winetricks
}

create_swtor_prefix() {
  log "(1/1) Creating Wine prefix"
  rm -rf "$PREFIX_DIR"
  WINEPREFIX="$PREFIX_DIR" wine wineboot
}

install_dll_vcrun2008() {
  log "(1/3) Installing vcrun2008"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q vcrun2008
}

install_dll_crypt32() {
  log "(2/3) Installing crypt32"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q crypt32
}

install_dll_d3dx9_36() {
  log "(3/3) Installing d3dx9_36"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q d3dx9_36
}

set_vram() {
  log "(1/3) Setting prefix VRAM to 1024"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q videomemorysize=1024
}

switch_windows_version() {
  log "(2/3) Switching Windows version to Windows 10"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q win10
}

switch_all_dlls_to_builtin() {
  log "(3/3) Change all prefix DLLs to be builtin"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q alldlls=builtin
}

download_swtor() {
  log "(1/1) Downloading SWTOR_setup.exe from http://www.swtor.com/download"
  wget -O "$PREFIX_DIR/drive_c/Program Files/SWTOR_setup.exe" $SWTOR_DOWNLOAD
}

launch_swtor() {
  log "Launching SWTOR_setup.exe..."
  WINEPREFIX="$PREFIX_DIR" wine "$PREFIX_DIR/drive_c/Program Files/SWTOR_setup.exe" >/dev/null 2>&1
}

build() {

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 1: Install Homebrew packages"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  install_package_wget
  tap_into_gcenx_brew
  install_package_wine_stable
  install_package_winetricks

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 2: Create custom Wine prefix"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  create_swtor_prefix

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 3: Install DLLs to prefix"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  # export WINE=wine32on64 # required to fool winetricks into using wine32on64
  install_dll_vcrun2008
  install_dll_crypt32
  install_dll_d3dx9_36

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 4: Change prefix settings"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  set_vram
  switch_windows_version
  switch_all_dlls_to_builtin

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 5: Download SWTOR executable"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  download_swtor

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 6: Launch SWTOR installer"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  launch_swtor
}

# Check if Homebrew is installed
if [[ $(command -v brew) == "" ]]; then
  log "ERROR: Homebrew not installed. Exiting."
else
  build
fi


package.sh
#!/usr/bin/env arch -x86_64 bash

set -e

SCRIPT_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
DIST_DIR="$SCRIPT_DIR/dist"
APP_DIR="$DIST_DIR/SWTOR.app"
PACKAGE_DIR="$DIST_DIR/SWTOR.dmg"
PREFIX_DIR="$SCRIPT_DIR/prefix"
TEMPLATE_DIR="$SCRIPT_DIR/template"

log() {
  echo -e "\033[01;35m\t$1\033[00m"
}

create_dist_dir() {
  log "->\tCreating dist dir"
  rm -rf "$DIST_DIR"
  mkdir -v "$DIST_DIR"
}

create_skeleton_app() {
  log "->\tCreating skeleton SWTOR.app"
  mkdir -vp "$APP_DIR"/Contents/{MacOS,Resources}

  cp -v "$TEMPLATE_DIR/Info.plist" "$APP_DIR/Contents/"
  cp -v "$TEMPLATE_DIR/SWTOR.icns" "$APP_DIR/Contents/Resources/"
  cp -v "$TEMPLATE_DIR/SWTOR.sh" "$APP_DIR/Contents/MacOS/SWTOR"
}

move_prefix_to_app() {
  log "->\tMoving prefix to SWTOR.app"
  mv -v "$PREFIX_DIR" "$APP_DIR/Contents/Resources/"
}

compress_app() {
  log "->\Compressing SWTOR.app as SWTOR.dmg"
  create-dmg --app-drop-link 580 110 --window-size 710 300 --icon SWTOR.app 120 110 --background "$TEMPLATE_DIR/background.png" $PACKAGE_DIR $APP_DIR
}

copy_wine_to_app() {
  log "->\Copying wine from Wine Crossover to SWTOR.app"
  cp -Rv "/Applications/Wine Crossover.app/Contents/Resources/wine" "$APP_DIR/Contents/Resources/"
}

move_back_prefix_dir() {
  log "->\tMoving prefix dir back to original place"
  mv -v "$APP_DIR/Contents/Resources/prefix" "$PREFIX_DIR"
}

remove_skeleton_app() {
  log "->\tRemoving skeleton SWTOR.app"
  rm -rf "$APP_DIR"
}

package() {

  log "_ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 1: Create dist dir"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  create_dist_dir

  log "_ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 2: Create SWTOR.app"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  create_skeleton_app
  move_prefix_to_app

  log "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _"
  log "Step 3: Prepare SWTOR package"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  copy_wine_to_app
  compress_app

  log "_ _ _ _ _ _ _ _"
  log "Step 9: Cleanup"
  log "‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾"

  move_back_prefix_dir
  remove_skeleton_app

}

if [[ -d "$PREFIX_DIR" ]]; then
  package
else
  log "ERROR: Prefix dir is not generated, please run build.sh first."
fi


shortcut/appify.sh
#!/usr/bin/env bash

VERSION=4.0.1
SCRIPT=`basename "$0"`
APPNAME="My App"
APPICONS="/System/Library/CoreServices/CoreTypes.bundle/Contents/Resources/GenericApplicationIcon.icns"
OSX_VERSION=`sw_vers -productVersion`
PWD=`pwd`

function usage {
	cat <<EOF

$SCRIPT v${VERSION} for for Mac OS X - https://gist.github.com/oubiwann/453744744da1141ccc542ff75b47e0cf

Usage:

  $SCRIPT [options]

Options:

  -h, --help 		Prints this help message, then exits
  -s, --script		Name of the script to 'appify' (required)
  -n, --name 		Name of the application (default "$APPNAME")
  -i, --icons		Name of the icons file to use when creating the app
                        (defaults to $APPICONS)
  -v, --version		Prints the version of this script, then exits

Description:

  Creates the simplest possible Mac app from a shell script.

  Appify has one required parameter, the script to appify:

    $SCRIPT --script my-app-script.sh

  Note that you cannot rename appified apps. If you want to give your app
  a custom name, use the '--name' option

    $SCRIPT --script my-app-script.sh --name "Sweet"

Copyright:

  Copyright (c) Thomas Aylott <http://subtlegradient.com/>
  Modified by Mathias Bynens <http://mathiasbynens.be/>
  Modified by Andrew Dvorak <http://OhReally.net/>
  Rewritten by Duncan McGreggor <http://github.com/oubiwann/>

EOF
	exit 1
}

function version {
	echo "v${VERSION}"
	exit 1
}

function error {
	echo
	echo "ERROR: $1"
	echo
	usage
}

while :; do
  case $1 in
    -h | --help )    usage;;
    -s | --script )  APPSCRIPT="$2"; shift ;;
    -n | --name )    APPNAME="$2"; shift ;;
    -i | --icons )   APPICONS="$2"; shift ;;
    -v | --version ) version;;
    -- )             shift; break ;;
    * )              break ;;
  esac
  shift
done

if [ -z ${APPSCRIPT+nil} ]; then
	error "the script to appify must be provided!"
fi

if [ ! -f "$APPSCRIPT" ]; then
	error "the can't find the script '$APPSCRIPT'"
fi

if [ -a "$APPNAME.app" ]; then
	error "the bundle '$PWD/$APPNAME.app' already exists"
fi

APPDIR="$APPNAME.app/Contents"

mkdir -vp "$APPDIR"/{MacOS,Resources}
cp -v "$APPICONS" "$APPDIR/Resources/$APPNAME.icns"
cp -v "$APPSCRIPT" "$APPDIR/MacOS/$APPNAME"
chmod +x "$APPDIR/MacOS/$APPNAME"

cat <<EOF > "$APPDIR/Info.plist"
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>CFBundleExecutable</key>
    <string>$APPNAME</string>
    <key>CFBundleGetInfoString</key>
    <string>$APPNAME</string>
    <key>CFBundleIconFile</key>
    <string>$APPNAME</string>
    <key>CFBundleName</key>
    <string>$APPNAME</string>
    <key>CFBundlePackageType</key>
    <string>APPL</string>
    <key>CFBundleSignature</key>
    <string>4242</string>
  </dict>
</plist>
EOF

echo "Application bundle created at '$PWD/$APPNAME.app'"
echo



patches/distversion.h
/* ---------------------------------------------------------------
 *   distversion.c
 *
 * Copyright 2013, CodeWeavers, Inc.
 * 
 * Information from DISTVERSION which needs to find 
 * its way into the wine tree.
 * --------------------------------------------------------------- */

#define WINDEBUG_WHAT_HAPPENED_MESSAGE "This can be caused by a problem in the program or a deficiency in Wine. You may want to check <a href=\"http://www.codeweavers.com/compatibility/\">http://www.codeweavers.com/compatibility/</a> for tips about running this application."

#define WINDEBUG_USER_SUGGESTION_MESSAGE "If this problem is not present under Windows and has not been reported yet, you can save the detailed information to a file using the \"Save As\" button, then <a href=\"http://www.codeweavers.com/support/tickets/enter/\">file a bug report</a> and attach that file to the report."

install.sh
#!/usr/bin/env arch -x86_64 bash

set -e

NONE='\033[00m'
PURPLE='\033[01;35m'
RED='\033[0;31m'

SCRIPT_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
PREFIX_DIR="$SCRIPT_DIR/prefix"

TOOLS_VERSION=$(xcode-select -p)
TOOLS_INSTALLED="/Library/Developer/CommandLineTools"
XCODE_CHECK=$(ls /Applications/Xcode.app) || : # set -e can cause the script to die here, so added || : to skip fail
XCODE_INSTALLED="Contents"
SWTOR_DOWNLOAD=http://www.swtor.com/download

install_package_wget() {
  echo -e "${PURPLE}\t(1/4) Installing wget${NONE}"
  brew install wget
}

tap_into_gcenx_brew() {
  echo -e "${PURPLE}\t(2/4) Tap into Gcenx/homebrew-wine${NONE}"
  brew tap Gcenx/homebrew-wine
}

install_package_wine_stable() {
  echo -e "${PURPLE}\t(3/4) Installing latest Wine version${NONE}"
  brew update
  brew install --cask --no-quarantine wine-crossover
}

install_package_winetricks() {
  echo -e "${PURPLE}\t(4/4) Installing Winetricks\n${NONE}"
  brew install winetricks
}

create_swtor_prefix() {
  echo -e "${PURPLE}\t(1/1) Creating Wine prefix\n${NONE}"
  rm -rf "$PREFIX_DIR"
  WINEARCH=win32 WINEPREFIX="$PREFIX_DIR" wine wineboot
}

install_dll_vcrun2008() {
  echo -e "${PURPLE}\t(1/3) Installing vcrun2008${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q vcrun2008
}

install_dll_crypt32() {
  echo -e "${PURPLE}\t(2/3) Installing crypt32${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q crypt32
}

install_dll_d3dx9_36() {
  echo -e "${PURPLE}\t(3/3) Installing d3dx9_36\n${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q d3dx9_36
}

set_vram() {
  echo -e "${PURPLE}\t(1/3) Setting prefix VRAM to 1024${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q videomemorysize=1024
}

switch_windows_version() {
  echo -e "${PURPLE}\t(2/3) Switching Windows version to Windows 10${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q win10
}

switch_all_dlls_to_builtin() {
  echo -e "${PURPLE}\t(3/3) Change all prefix DLLs to be builtin\n${NONE}"
  env WINEPREFIX="$PREFIX_DIR" sh winetricks -q alldlls=builtin
}

download_swtor() {
  echo -e "${PURPLE}\t(1/1) Downloading SWTOR_setup.exe from http://www.swtor.com/download${NONE}"
  wget -O "$PREFIX_DIR/drive_c/Program Files/SWTOR_setup.exe" $SWTOR_DOWNLOAD
}

launch_swtor() {
  echo -e "${PURPLE}\tLaunching SWTOR_setup.exe...${NONE}"
  WINEPREFIX="$PREFIX_DIR" wine "$PREFIX_DIR/drive_c/Program Files/SWTOR_setup.exe" >/dev/null 2>&1
}

# Main function that installs Homebrew packages and SWTOR
install() {

  echo -e "${PURPLE}\tStep 1: Install Homebrew packages${NONE}"
  echo -e "${PURPLE}\t‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ${NONE}"

  install_package_wget
  tap_into_gcenx_brew
  install_package_wine_stable
  install_package_winetricks

  echo -e "${PURPLE}\tStep 2: Create custom Wine prefix${NONE}"
  echo -e "${PURPLE}\t‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ${NONE}"

  create_swtor_prefix

  echo -e "${PURPLE}\tStep 3: Install DLLs to prefix${NONE}"
  echo -e "${PURPLE}\t‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ${NONE}"

  # export WINE=wine32on64 # required to fool winetricks into using wine32on64
  install_dll_vcrun2008
  install_dll_crypt32
  install_dll_d3dx9_36

  echo -e "${PURPLE}\tStep 4: Change prefix settings${NONE}"
  echo -e "${PURPLE}\t‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ${NONE}"

  set_vram
  switch_windows_version
  switch_all_dlls_to_builtin

  echo -e "${PURPLE}\tStep 5: Download SWTOR executable${NONE}"
  echo -e "${PURPLE}\t‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾ ‾${NONE}"

  download_swtor

  echo -e "${PURPLE}\tStep 6: Launch SWTOR installer${NONE}"

  launch_swtor
}

# Check if Command Line Tools are installed followed by if Homebrew is installed
# If either isn't installed, the script will quit
if [ "$TOOLS_VERSION" = "$TOOLS_INSTALLED" ] || [ "$XCODE_CHECK" = "$XCODE_INSTALLED" ]; then
  if [[ $(command -v brew) == "" ]]; then
    echo -e "${RED}\tERROR: Homebrew not installed. Exiting.${NONE}"
  else
    install
  fi
else
  echo -e "${RED}\tERROR: Command Line Tools not installed. Exiting.${NONE}"
fi
